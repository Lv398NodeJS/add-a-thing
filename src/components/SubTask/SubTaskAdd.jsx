import React, { Component } from 'react';
import { InputGroup, Button, FormControl } from 'react-bootstrap';

class SubTaskAdd extends Component {
  render() {
    const { addSubTask } = this.props;
    return (
      <InputGroup className="mb-3">
        <InputGroup.Prepend>
          <Button
            variant="outline-success"
            size="sm"
            onClick={() => {
              addSubTask(this.input.value);
              this.input.value = '';
            }}
          >
            +
          </Button>
        </InputGroup.Prepend>
        <FormControl
          size="sm"
          maxLength={40}
          ref={(input) => {
            this.input = input;
          }}
          onKeyPress={(event) => {
            if (event.key === 'Enter') {
              addSubTask(this.input.value);
              this.input.value = '';
            }
          }}
        />
      </InputGroup>
    );
  }
}

export default SubTaskAdd;
