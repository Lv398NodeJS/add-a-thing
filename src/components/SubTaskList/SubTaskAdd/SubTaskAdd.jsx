import React, { Component } from 'react';
import {
  Row, Col, InputGroup, Button, Form,
} from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as subTaskActions from '../../../actions/subTaskListActions';

class SubTaskAdd extends Component {
  handleSubmit = (event) => {
    const { taskRef, subTaskListActions } = this.props;
    event.preventDefault();
    event.stopPropagation();
    const text = this.input.value.trim().replace(/\s+/g, ' ');
    if (this.form.checkValidity() === false) {
      this.form.classList.add('was-validated');
    } else {
      this.form.classList.remove('was-validated');
      subTaskListActions.addSubTask(taskRef, text);
      this.input.value = '';
    }
  };

  render() {
    const { taskStatus } = this.props;

    return (
      <Row className="mb-0 mt-3 mx-0">
        <Col className="px-0">
          <Form
            className="needs-validation"
            method="POST"
            noValidate
            onSubmit={event => this.handleSubmit(event)}
            ref={(form) => {
              this.form = form;
            }}
          >
            <InputGroup>
              <InputGroup.Prepend>
                <Button
                  className="add-subtask-button"
                  variant="outline-success"
                  size="sm"
                  as="input"
                  type="submit"
                  value="+"
                  disabled={taskStatus === 'Done'}
                />
              </InputGroup.Prepend>
              <Form.Control
                className="new-subtask-text was-validated"
                size="sm"
                minLength={1}
                maxLength={40}
                placeholder="Enter subtask text here..."
                disabled={taskStatus === 'Done'}
                required
                ref={(input) => {
                  this.input = input;
                }}
                onChange={(event) => {
                  if (!event.target.value.trim().length) event.target.value = '';
                }}
              />
              <div className="valid-feedback">Looks good!</div>
              <div className="invalid-feedback">Please provide subtask text.</div>
            </InputGroup>
          </Form>
        </Col>
      </Row>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  subTaskListActions: bindActionCreators(subTaskActions, dispatch),
});
export { SubTaskAdd as SubTaskAddComponent };
export default connect(
  null,
  mapDispatchToProps,
)(SubTaskAdd);
