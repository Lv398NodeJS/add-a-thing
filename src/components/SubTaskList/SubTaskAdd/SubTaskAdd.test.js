import React from 'react';
import { shallow } from 'enzyme';
import SubTaskAdd from './SubTaskAdd';

describe('SubTaskAdd component', () => {
  it('should render correctly with no props', () => {
    const subTaskAddComponent = shallow(<SubTaskAdd />);
    expect(subTaskAddComponent).toMatchSnapshot();
  });

  it('should always render button with ".add-subtask-button" class', () => {
    const subTaskAddComponent = shallow(<SubTaskAdd />);
    const button = subTaskAddComponent.find('.add-subtask-button');
    expect(button.length).toBeGreaterThan(0);
  });

  it('should always render text input with ".new-subtask-text" class', () => {
    const subTaskAddComponent = shallow(<SubTaskAdd />);
    const input = subTaskAddComponent.find('.new-subtask-text');
    expect(input.length).toBeGreaterThan(0);
  });
});
