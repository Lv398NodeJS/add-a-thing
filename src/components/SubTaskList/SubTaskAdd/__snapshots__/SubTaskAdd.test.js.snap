// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubTaskAdd component should render correctly and match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SubTaskAdd />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <ForwardRef
        as="div"
        className="px-0"
      >
        <ForwardRef(render)>
          <ForwardRef>
            <ForwardRef
              active={false}
              as="input"
              className="add-subtask-button"
              disabled={false}
              onClick={[Function]}
              size="sm"
              type="button"
              value="+"
              variant="outline-success"
            />
          </ForwardRef>
          <ForwardRef
            as="input"
            className="new-subtask-text"
            disabled={false}
            maxLength={40}
            minLength={1}
            onKeyPress={[Function]}
            placeholder="Enter subtask text here..."
            size="sm"
          />
        </ForwardRef(render)>
      </ForwardRef>,
      "className": "mb-0 mt-3 mx-0",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "as": "div",
        "children": <ForwardRef(render)>
          <ForwardRef>
            <ForwardRef
              active={false}
              as="input"
              className="add-subtask-button"
              disabled={false}
              onClick={[Function]}
              size="sm"
              type="button"
              value="+"
              variant="outline-success"
            />
          </ForwardRef>
          <ForwardRef
            as="input"
            className="new-subtask-text"
            disabled={false}
            maxLength={40}
            minLength={1}
            onKeyPress={[Function]}
            placeholder="Enter subtask text here..."
            size="sm"
          />
        </ForwardRef(render)>,
        "className": "px-0",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <ForwardRef>
              <ForwardRef
                active={false}
                as="input"
                className="add-subtask-button"
                disabled={false}
                onClick={[Function]}
                size="sm"
                type="button"
                value="+"
                variant="outline-success"
              />
            </ForwardRef>,
            <ForwardRef
              as="input"
              className="new-subtask-text"
              disabled={false}
              maxLength={40}
              minLength={1}
              onKeyPress={[Function]}
              placeholder="Enter subtask text here..."
              size="sm"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <ForwardRef
                active={false}
                as="input"
                className="add-subtask-button"
                disabled={false}
                onClick={[Function]}
                size="sm"
                type="button"
                value="+"
                variant="outline-success"
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "active": false,
                "as": "input",
                "className": "add-subtask-button",
                "disabled": false,
                "onClick": [Function],
                "size": "sm",
                "type": "button",
                "value": "+",
                "variant": "outline-success",
              },
              "ref": null,
              "rendered": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "active": false,
                  "disabled": false,
                  "type": "button",
                  "variant": "primary",
                },
                "displayName": "Button",
                "render": [Function],
              },
            },
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "displayName": "InputGroupPrepend",
              "render": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": "input",
              "className": "new-subtask-text",
              "disabled": false,
              "maxLength": 40,
              "minLength": 1,
              "onKeyPress": [Function],
              "placeholder": "Enter subtask text here...",
              "size": "sm",
            },
            "ref": [Function],
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "Feedback": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "as": "div",
                  "type": "valid",
                },
                "displayName": "Feedback",
                "propTypes": Object {
                  "as": [Function],
                  "type": [Function],
                },
                "render": [Function],
              },
              "defaultProps": Object {
                "as": "input",
              },
              "displayName": "FormControl",
              "render": [Function],
            },
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "Append": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "displayName": "InputGroupAppend",
            "render": [Function],
          },
          "Checkbox": [Function],
          "Prepend": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "displayName": "InputGroupPrepend",
            "render": [Function],
          },
          "Radio": [Function],
          "Text": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "displayName": "InputGroupText",
            "render": [Function],
          },
          "defaultProps": undefined,
          "displayName": "Bootstrap(InputGroup)",
          "propTypes": undefined,
          "render": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "as": "div",
        },
        "displayName": "Col",
        "render": [Function],
      },
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": undefined,
      "displayName": "Bootstrap(Row)",
      "propTypes": undefined,
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <ForwardRef
          as="div"
          className="px-0"
        >
          <ForwardRef(render)>
            <ForwardRef>
              <ForwardRef
                active={false}
                as="input"
                className="add-subtask-button"
                disabled={false}
                onClick={[Function]}
                size="sm"
                type="button"
                value="+"
                variant="outline-success"
              />
            </ForwardRef>
            <ForwardRef
              as="input"
              className="new-subtask-text"
              disabled={false}
              maxLength={40}
              minLength={1}
              onKeyPress={[Function]}
              placeholder="Enter subtask text here..."
              size="sm"
            />
          </ForwardRef(render)>
        </ForwardRef>,
        "className": "mb-0 mt-3 mx-0",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "div",
          "children": <ForwardRef(render)>
            <ForwardRef>
              <ForwardRef
                active={false}
                as="input"
                className="add-subtask-button"
                disabled={false}
                onClick={[Function]}
                size="sm"
                type="button"
                value="+"
                variant="outline-success"
              />
            </ForwardRef>
            <ForwardRef
              as="input"
              className="new-subtask-text"
              disabled={false}
              maxLength={40}
              minLength={1}
              onKeyPress={[Function]}
              placeholder="Enter subtask text here..."
              size="sm"
            />
          </ForwardRef(render)>,
          "className": "px-0",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <ForwardRef>
                <ForwardRef
                  active={false}
                  as="input"
                  className="add-subtask-button"
                  disabled={false}
                  onClick={[Function]}
                  size="sm"
                  type="button"
                  value="+"
                  variant="outline-success"
                />
              </ForwardRef>,
              <ForwardRef
                as="input"
                className="new-subtask-text"
                disabled={false}
                maxLength={40}
                minLength={1}
                onKeyPress={[Function]}
                placeholder="Enter subtask text here..."
                size="sm"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <ForwardRef
                  active={false}
                  as="input"
                  className="add-subtask-button"
                  disabled={false}
                  onClick={[Function]}
                  size="sm"
                  type="button"
                  value="+"
                  variant="outline-success"
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "active": false,
                  "as": "input",
                  "className": "add-subtask-button",
                  "disabled": false,
                  "onClick": [Function],
                  "size": "sm",
                  "type": "button",
                  "value": "+",
                  "variant": "outline-success",
                },
                "ref": null,
                "rendered": null,
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": Object {
                    "active": false,
                    "disabled": false,
                    "type": "button",
                    "variant": "primary",
                  },
                  "displayName": "Button",
                  "render": [Function],
                },
              },
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "displayName": "InputGroupPrepend",
                "render": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "as": "input",
                "className": "new-subtask-text",
                "disabled": false,
                "maxLength": 40,
                "minLength": 1,
                "onKeyPress": [Function],
                "placeholder": "Enter subtask text here...",
                "size": "sm",
              },
              "ref": [Function],
              "rendered": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "Feedback": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": Object {
                    "as": "div",
                    "type": "valid",
                  },
                  "displayName": "Feedback",
                  "propTypes": Object {
                    "as": [Function],
                    "type": [Function],
                  },
                  "render": [Function],
                },
                "defaultProps": Object {
                  "as": "input",
                },
                "displayName": "FormControl",
                "render": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Append": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "displayName": "InputGroupAppend",
              "render": [Function],
            },
            "Checkbox": [Function],
            "Prepend": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "displayName": "InputGroupPrepend",
              "render": [Function],
            },
            "Radio": [Function],
            "Text": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "displayName": "InputGroupText",
              "render": [Function],
            },
            "defaultProps": undefined,
            "displayName": "Bootstrap(InputGroup)",
            "propTypes": undefined,
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "as": "div",
          },
          "displayName": "Col",
          "render": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "displayName": "Bootstrap(Row)",
        "propTypes": undefined,
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
